### ISSUE TITLE
Use name of wiki for interwiki links

### ISSUE THREAD

**Issue by davidlmontgomery** [vimwiki/vimwiki#477]
TLDR: Enhancement proposal: allow wikis to have names, and support using these names in links.

I have found vimwiki's interwiki link schema ("wiki" + number of wiki +  ":") to be inconvenient. For one thing, the numbers are unnecessarily hard to remember, especially since the values differ from those used with VimwikiUISelect. This makes creating the links correctly more difficult, and it makes the links more difficult to understand. "Which one is wiki7?" 

But the biggest thing is that they are brittle. If I add a wiki to my wiki list, I break all my interwiki links that point to wikis later in the list. Same if I remove a wiki. Or if I reorder my wikis. Using the schema with a number means that any changes to g:vimwiki_list may break an arbitrary number of interwiki links. 

The solution I propose is to allow users to provide a name for each wiki, if they desire, and then to allow use of that name in an interwiki link syntax. Essentially, this is like DNS. Each wiki has a numeric address, yes, but for human-computer interaction using a meaningful (domain or wiki) name is usually superior. It is more understandable, and it provides a level of indirection allowing the underlying numeric address to change without needing to fix up any links.

At first I just implemented this using VimwikiLinkHandler:

```
" VimwikiHandler for named interwiki links 
function! VimwikiLinkHandler(cmd, link)
  " Use Vim to open named interwiki files via the 'wn.<name>:' scheme.  
  " For example [[wn.spanish:cognates]] 
  let lnk = a:link
  let wnscheme = '^wn\.\([a-zA-Z0-9]\+\):\(.*\)' 
  let wnmatch = matchlist(lnk, wnscheme)
  if len(wnmatch) >= 2 && wnmatch[1] != '' && wnmatch[2] != '' 
    let wname = wnmatch[1]
    let idx = 0
    while idx < len(g:vimwiki_list)
      let wiki = g:vimwiki_list[idx] 
      if has_key(wiki, 'name')
        if wiki.name == wname 
          let lnk = 'wiki' . idx . ':' . wnmatch[2] 
          let lnk = substitute(lnk, VimwikiGet('ext', idx).'$', '', '')
          call vimwiki#base#open_link(a:cmd, lnk)
          return 1
        endif
      endif
      let idx += 1
    endwhile
    echom 'Vimwiki Error: No wiki found with name "' . wiki . '"' 
    return 0
  endif
  return 0
endfunction  
```

Although a user can do this with VimwikiLinkHandler, I feel that this kind of functionality would be useful out of the box, so ultimately I decided to integrate this into base.vim as shown here https://github.com/davidlmontgomery/vimwiki/commit/4356cd3fb49df0c474bf5ace5e7fa20245582183 .
 
Once we have names for wikis, there are other ways to take advantage of them. For example, I now show the wiki names (when available) for VimwikiUISelect, using the code here https://github.com/davidlmontgomery/vimwiki/commit/8447c3ce58a1f50e41bcb9d0da252f7c5a7e0ebe

(By the way, my patches are relative to 2.3, because that was what vim-plug deployed for me. I see now that perhaps I should be running off the tip of the dev branch. I have the intention of switching to that when I have time to test it out. After that I'll be able to provide patches that apply more directly.)

**Comment by EinfachToll** [comment]
This is actually a duplicate of #32, but I will close that because this here is a bit more detailed.
If want to provide patches, you should definitely start from the dev branch because there were some large code refactorings in the meantime.

**Comment by ranebrown** [comment]
@davidlmontgomery If you create a PR against the `dev` branch I will review and merge. I think this would be a pretty useful feature.
